x/emissions/keeper/actor_penalties.go

```diff
@@ -0,0 +1,126 @@
+	if err != nil {
+		return errorsmod.Wrap(err, "error getting lowest inferer score EMA")
+	}
+
+	// Empty active set so no penalty
+	if !exists {
+		err = k.penaltyInTopic.Remove(ctx, topicId)
+		if err != nil {
+			return errorsmod.Wrap(err, "error removing penalty for topic")
+		}
+	}
+
+	// Get the EMA score from the active set
+	var scores []alloraMath.Dec
+	if err := k.activeInferers.Walk(
+		ctx,
+		collections.NewPrefixedPairRange[TopicId, ActorId](topicId),
+		func(key collections.Pair[TopicId, ActorId]) (stop bool, err error) {
+			score, err := k.GetInfererScoreEma(ctx, topicId, key.K2())
+			if err != nil {
+				return false, errorsmod.Wrap(err, "error getting inferer score EMA")
+			}
+			scores = append(scores, score.Score)
+			return false, nil
+		},
+	); err != nil {
+		return errorsmod.Wrap(err, "error walking active inferers")
+	}
+
+	params, err := k.GetParams(ctx)
+	if err != nil {
+		return errorsmod.Wrap(err, "error getting params")
+	}
+	standardDeviation, _, err := alloraMath.MedianAbsoluteDeviation(scores)
+	if err != nil {
+		return errorsmod.Wrap(err, "error calculating median absolute deviation")
+	}
+	x, err := standardDeviation.Mul(params.SortitionLambdaPenalty)
+	if err != nil {
+		return errorsmod.Wrap(err, "error calculating penalty")
+	}
+	penalty, err := lowestEMAScore.Score.Sub(x)
+	if err != nil {
+		return errorsmod.Wrap(err, "error calculating penalty")
+	}
+
+	return k.penaltyInTopic.Set(ctx, topicId, penalty)
+}
+
+// MayPenaliseInferer penalises an inferer for missing previous epochs. It saves and returns the new EMA score.
+// If the inferer didn't miss any epochs this is a no-op, the EMA score is returned as is.
+func (k *Keeper) MayPenaliseInferer(
+	ctx sdk.Context,
+	topic types.Topic,
+	block types.BlockHeight,
+	emaScore types.Score,
+) (types.Score, error) {
+	missedEpochs := countWorkerContiguousMissedEpochs(topic, emaScore.BlockHeight)
+	// No missed epochs == no penalty
+	if missedEpochs == 0 {
+		return emaScore, nil
+	}
+
+	penalty, err := k.penaltyInTopic.Get(ctx, topic.Id)
+	// No penalty set, no penalty
+	if errors.Is(err, collections.ErrNotFound) {
+		return emaScore, nil
+	}
+	if err != nil {
+		return types.Score{}, err
+	}
+	emaScore.BlockHeight = block
+	// TODO: Provide the limit as the initial score
+	emaScore.Score, err = applyPenalty(topic, penalty, alloraMath.ZeroDec(), emaScore.Score, missedEpochs)
+	if err != nil {
+		return types.Score{}, err
+	}
+
+	// Save the penalised EMA score
+	return emaScore, k.SetInfererScoreEma(ctx, topic.Id, emaScore.Address, emaScore)
+}
+
+// applyPenalty applies the penalty to the EMA score for the given number of missed epochs while staying above provided limit.
+func applyPenalty(topic types.Topic, penalty, limit, emaScore alloraMath.Dec, missedEpochs uint32) (alloraMath.Dec, error) {
+	for i := uint32(0); i < missedEpochs; i++ {
+		penalisedScore, err := alloraMath.CalcEma(topic.MeritSortitionAlpha, penalty, emaScore, false)
```

Comment:
This way of calculating feels like very cumbersome... 
0/ Is this one the one agreed as the way to apply an epoch-based loop penalty?
1/ Should we set a limit of max penalties? Just to try to not get too many epochs (=unbounded) which could affect block stability. Should be agreed though. (This is a common thing we want to avoid in general, worth having at the back of our minds lol)
2/ Could there be an alternative way to compute this penalty?

Reply:
We could set a maximum number of `missedEpoch` to consider

Reply:
a limit set on emissions params, yep

Reply:
It's been replaced by a new formula so no more loop :)